src/ir/stack-utils.cpp

```diff
@@ -160,6 +160,42 @@ StackSignature StackSignature::operator+(const StackSignature& next) const {
   return sig;
 }
 
+StackSignature StackSignature::getLeastUpperBound(StackSignature a,
+                                                  StackSignature b) {
+  // Assert an approximation of the conditions for having a LUB
+  assert((a.params.size() >= b.params.size() &&
+          a.results.size() >= b.results.size()) ||
+         a.unreachable);
+  assert((b.params.size() >= a.params.size() &&
+          b.results.size() >= a.results.size()) ||
+         b.unreachable);
+  // Canonicalize so that the shorter types are on `a`.
+  if (a.params.size() > b.params.size()) {
+    std::swap(a.params, b.params);
+  }
+  if (a.results.size() > b.results.size()) {
+    std::swap(a.results, b.results);
+  }
+  std::vector<Type> params;
+  for (size_t i = 0, size = a.params.size(); i < size; ++i) {
+    Type lub = Type::getLeastUpperBound(a.params[i], b.params[i]);
+    assert(lub != Type::none);
+    params.push_back(lub);
+  }
+  params.insert(
+    params.end(), b.params.begin() + a.params.size(), b.params.end());
+  size_t resultsDiff = b.results.size() - a.results.size();
+  std::vector<Type> results(b.results.begin(), b.results.begin() + resultsDiff);
+  for (size_t i = 0, size = a.results.size(); i < size; ++i) {
+    Type lub =
+      Type::getLeastUpperBound(a.results[i], b.results[i + resultsDiff]);
+    assert(lub != Type::none);
+    results.push_back(lub);
+  }
+  return StackSignature{
+    Type(params), Type(results), a.unreachable && b.unreachable};
```

Comment:
Being unreachable means that more elements can be appended to the params or prepended to the results as necessary. If two stack signatures have this property, then their LUB would have this property as well. If either signature was not unreachable and could not be extended, then the LUB should not be able to be extended either. This would be important for calculating the LUB of more than two unreachable stack signatures so that the number of params and results is not "frozen" by whatever the first two signatures to be combined are.

Reply:
I'm still not sure if I understand the concept of LUB here. In the usual typing sense LUB of two types is a supertype of the two but the lowest one in the type hierarchy. First I'm not sure how unreachability is considered in this type hierarchy, I could be wrong, but I feel, for the two signatures with the same type input and output [a]->[b], if one of them is unreachable, that means that type has more flexibility and it should be considered as a supertype of the other.

But here it is possible that we have a LUB of two types that is not unreachable but one of those two child types can be unreachable. I'm not sure if I explained my confusion very well, but I'd appreciate a bit more explanation, maybe with an example.

Reply:
Here's the subtype relation on stack signatures, which answers the question "can an instruction sequence with the LHS stack signature be typed with the RHS stack signature?"

`[t1*] -> [t2*] <: [s* t1'*] -> [s* t2'*]` iff

 - `t1'_i <: t1_i`
 -  `t2_i <: t2'_i`

`[t1*] -> [t2*] {u} <: [s1* t1'*] -> [s2* t2'*] {u?}` iff

 - `[t1*] -> [t2*] <: [t1'*] -> [t2'*]`

As an example of the first rule, consider the instruction sequence

```
ref.as_func
drop
i32.add
```

The most specific type you could give this sequence is `[i32, i32, anyref] -> [i32]`. But you could also use it in a context that expects `[i32, i32, funcref] -> [i32]` because `ref.as_func` can accept `funcref` or any other subtype of `anyref`. That's where the contravariance comes from. You could also use this instruction sequence anywhere that expects `[f32, i32, i32, anyref] -> [f32, anyref]` because the `f32` simply stays on the stack throughout the sequence. That's where the the prefix extension comes from.

For the second rule, consider this sequence:

```
ref.as_func
drop
i32.add
unreachable
i32.const 0
```

This instruction sequence has the specific type `[i32, i32, anyref] -> [i32] {u}`. It can be used in any situation the previous block can be used in, but can additionally be used in contexts that expect something like `[f32, i32, i32, anyref] -> [v128, i32]`. Because of the unreachable polymorphism, the additional prefixes on the params and results do not need to match.

Note that a reachable stack signature (without a `{u}`) is never a subtype of any unreachable stack signature (with a {u}). This makes sense because a sequence of instructions that has no polymorphic unreachable behavior cannot be given a type that says it does have polymorphic unreachable behavior.

Also, `[] -> [] {u}` is the bottom type here; it is a subtype of every other stack signature. This corresponds to `(unreachable)` being able to be given any stack signature.

Having written this out, I realize the implementation does not correctly handle the contravariance of the params. I will fix that, although it won't matter until we support block inputs.